src/syntaxtree.c

```diff
@@ -0,0 +1,898 @@
+
+    return result;
+}
+
+Obj SyntaxTreePermExpr(Obj result, Expr expr)
+{
+    Obj  cycles;
+    Obj  cycle;
+    Obj  val;
+    Expr cycleexpr;
+    Int  csize, n;
+    Int  i, j;
+
+    /* determine number of cycles */
+    n = SIZE_EXPR(expr) / sizeof(Expr);
+    cycles = NEW_PLIST(T_PLIST, n);
+    AssPRec(result, RNamName("cycles"), cycles);
+    SET_LEN_PLIST(cycles, n);
+
+    /* enter cycles */
+    for (i = 1; i <= n; i++) {
+        cycleexpr = ADDR_EXPR(expr)[i - 1];
+        csize = SIZE_EXPR(cycleexpr) / sizeof(Expr);
+        cycle = NEW_PLIST(T_PLIST, csize);
+        SET_LEN_PLIST(cycle, csize);
+        SET_ELM_PLIST(cycles, i, cycle);
+        CHANGED_BAG(cycles);
+
+        /* entries of the cycle */
+        for (j = 1; j <= csize; j++) {
+            val = SyntaxTreeCompiler(ADDR_EXPR(cycleexpr)[j - 1]);
+            SET_ELM_PLIST(cycle, j, val);
+            CHANGED_BAG(cycle);
+        }
+    }
+    return result;
+}
+
+Obj SyntaxTreeListExpr(Obj result, Expr expr)
+{
+    Obj list;
+    Int len;
+    Int i;
+
+    len = SIZE_EXPR(expr) / sizeof(Expr);
+
+    list = NEW_PLIST(T_PLIST, len);
+    SET_LEN_PLIST(list, len);
+
+    for (i = 1; i <= len; i++) {
+        if (ADDR_EXPR(expr)[i - 1] == 0) {
+            continue;
+        }
+        else {
+            SET_ELM_PLIST(list, i,
+                          SyntaxTreeCompiler(ADDR_EXPR(expr)[i - 1]));
+            CHANGED_BAG(list);
+        }
+    }
+
+    AssPRec(result, RNamName("list"), list);
+
+    return result;
+}
+
+Obj SyntaxTreeRangeExpr(Obj result, Expr expr)
+{
+    Obj first;
+    Obj second;
+    Obj last;
+
+    if (SIZE_EXPR(expr) == 2 * sizeof(Expr)) {
+        first = SyntaxTreeCompiler(ADDR_EXPR(expr)[0]);
+        last = SyntaxTreeCompiler(ADDR_EXPR(expr)[1]);
+
+        AssPRec(result, RNamName("first"), first);
+        AssPRec(result, RNamName("last"), last);
+    }
+    else {
+        first = SyntaxTreeCompiler(ADDR_EXPR(expr)[0]);
+        second = SyntaxTreeCompiler(ADDR_EXPR(expr)[1]);
+        last = SyntaxTreeCompiler(ADDR_EXPR(expr)[2]);
+
+        AssPRec(result, RNamName("first"), first);
+        AssPRec(result, RNamName("second"), second);
+        AssPRec(result, RNamName("last"), last);
+    }
+
+    return result;
+}
+
+Obj SyntaxTreeStringExpr(Obj result, Expr expr)
+{
+    Obj string;
+
+    C_NEW_STRING(string, SIZE_EXPR(expr) - 1 - sizeof(UInt),
+                 sizeof(UInt) + (Char *)ADDR_EXPR(expr));
+
+    AssPRec(result, RNamName("string"), string);
+
+    return result;
+}
+
+/* TODO: Maybe return a record or a list with key/value pairs? */
+Obj SyntaxTreeRecExpr(Obj result, Expr expr)
+{
+    Obj  key;
+    Obj  val;
+    Obj  list;
+    Obj  subrec;
+    Expr tmp;
+    Int  i, len;
+
+    len = SIZE_EXPR(expr) / (2 * sizeof(Expr));
+    list = NEW_PLIST(T_PLIST, len);
+    SET_LEN_PLIST(list, len);
+
+    for (i = 1; i <= len; i++) {
+        // Check whether this can happen and *why*
```

Comment:
```suggestion
        // TODO: check whether this can happen and *why*
```

Reply:
I am not quite sure what this comment refers to, though. Perhaps the `tmp == 0` case? Then perhaps it should be moved into that, to make it clearer?

But then, looking at the coder, and also at `RecExpr2`, I am pretty sure `tmp == 0` can never happen, so I'd replaced this by `GAP_ASSERT(tmp != 0);` and be done with it.